utils-intervalrdd/src/main/scala/org/bdgenomics/utils/interval/array/IntervalArray.scala

```diff
@@ -144,6 +160,49 @@ trait IntervalArray[K <: Interval[K], T] extends Serializable {
     }
   }
 
+  /**
+   * Recursively finds an index in array that overlaps with the query rr, or if
+   * none exists returns the closest non-overlapping index.
+   *
+   * @see binarySearch
+   *
+   * @param rr the query Interval.
+   * @param idx the current index to search.
+   * @param step the step size of the current search in array.
+   * @return the index of array that overlaps or is closest to query Interval rr.
+   */
+  @tailrec private def binaryNearestSearch(rr: K,
+                                           idx: Int = 0,
+                                           step: Int = midpoint): Int = {
+
+    if (array.length == 0 || rr.covers(array(idx)._1) || step == 0) {
+      idx
+    } else {
+      val stepIdx = idx + step
+      val nextIdx = if (stepIdx >= length ||
+        (!rr.covers(array(stepIdx)._1) &&
+          rr.compareTo(array(stepIdx)._1) < 0)) {
+        idx
+      } else {
+        stepIdx
+      }
+      binaryNearestSearch(rr, nextIdx, step / 2)
+    }
+  }
+
+  /**
+   * Recursively expands the list containing all Intervals in array that overlap
+   * with the query rr. Expands the list forward from the given index.
+   *
+   * @see expandBackward
+   *
+   * @param rr the query Interval.
+   * @param idx the index in array to begin expanding from.
+   * @param list the current list of tuples from array that overlap with query
```

Comment:
...with query Interval rr. &rarr; ...with the query interval.

And similar below.